File: solutions/javascript/src/structure/dynamic/DynamicPriorityQueueADT.js
class PriorityQueueNode {
  constructor(value, priority) {
    this.value = value;
    this.priority = priority;
    this.next = null;
  }
}

class DynamicPriorityQueueADT {
  constructor() {
    this.front = null;
  }

  // Agrega un elemento respetando el orden de prioridad (menor número = mayor prioridad)
  add(value, priority) {
    const newNode = new PriorityQueueNode(value, priority);
    if (this.isEmpty() || priority < this.front.priority) {
      newNode.next = this.front;
      this.front = newNode;
    } else {
      let current = this.front;
      while (current.next !== null && current.next.priority <= priority) {
        current = current.next;
      }
      newNode.next = current.next;
      current.next = newNode;
    }
  }

  // Devuelve el valor del primer elemento (con mayor prioridad)
  getElement() {
    if (this.isEmpty()) throw new Error("La cola está vacía");
    return this.front.value;
  }

  // Devuelve la prioridad del primer elemento
  getPriority() {
    if (this.isEmpty()) throw new Error("La cola está vacía");
    return this.front.priority;
  }

  // Elimina el primer elemento (con mayor prioridad)
  remove() {
    if (this.isEmpty()) throw new Error("La cola está vacía");
    this.front = this.front.next;
  }

  // Devuelve true si no hay elementos
  isEmpty() {
    return this.front === null;
  }
}

export default DynamicPriorityQueueADT;

File: solutions/python/structure/dynamic/DynamicPriorityQueueADT.py
class PriorityNode:
    def __init__(self, value: int, priority: int, siguiente=None):
        self.value = value
        self.priority = priority
        self.siguiente = siguiente

class DynamicPriorityQueueADT:
    def __init__(self):
        self.node = None

    def get_element(self) -> int:
        if self.is_empty():
            raise Exception("No se puede obtener el primero de una cola vacía")
        return self.node.value

    def get_priority(self) -> int:
        if self.is_empty():
            raise Exception("No se puede obtener la prioridad del primero de una cola vacía")
        return self.node.priority

    def remove(self):
        if self.is_empty():
            raise Exception("No se puede remover el primero de una cola vacía")
        self.node = self.node.siguiente

    def add(self, value: int, priority: int):
        nuevo = PriorityNode(value, priority)

        if self.is_empty() or self.node.priority > priority:
            nuevo.siguiente = self.node
            self.node = nuevo
            return

        actual = self.node
        while actual.siguiente is not None and actual.siguiente.priority <= priority:
            actual = actual.siguiente

        nuevo.siguiente = actual.siguiente
        actual.siguiente = nuevo

    def is_empty(self) -> bool:
        return self.node is None

    def mostrar(self):
        actual = self.node
        while actual:
            print(f"[{actual.value}, {actual.priority}]", end=" -> ")
            actual = actual.siguiente
        print("None")
